--- org/jivesoftware/smack/packet/UnknownPacket.java	2011-09-08 18:31:29.249404228 -0700
+++ org/jivesoftware/smack/packet/UnknownPacket.java	1969-12-31 16:00:00.000000000 -0800
@@ -0,0 +1,38 @@
+/**
+ * $RCSfile$
+ * $Revision: 11824 $
+ * $Date: 2010-08-15 08:13:05 -0700 (Sun, 15 Aug 2010) $
+ *
+ * Copyright 2003-2007 Jive Software.
+ *
+ * All rights reserved. Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jivesoftware.smack.packet;
+
+import org.jivesoftware.smack.util.StringUtils;
+
+
+/**
+ */
+public abstract class UnknownPacket extends Packet implements PacketExtension {
+	public String toXML() {
+		StringBuilder buf = new StringBuilder();
+		buf.append("<").append(getElementName());
+        if (getXmlns() != null) {
+            buf.append(" xmlns=\"").append(getNamespace()).append("\"");
+        }
+        buf.append("/>");
+        return buf.toString();
+	}
+}
\ No newline at end of file
--- org/jivesoftware/smack/XMPPConnection.java	2011-09-08 21:29:50.259401439 -0700
+++ org/jivesoftware/smack/XMPPConnection.java	2012-02-28 14:57:29.390749403 -0800
@@ -213,6 +213,9 @@
         }
 
         // Set the user.
+        if (resource == null)
+        	return ;
+        
         if (response != null) {
             this.user = response;
             // Update the serviceName with the one returned by the server
@@ -952,7 +949,11 @@
      *      appropiate error messages to end-users.
      */
     public void connect() throws XMPPException {
+    	connect(true);
+    }
+
+    public void connect(boolean bind) throws XMPPException {
+        // Establishes the connection, readers and writers
-        // Stablishes the connection, readers and writers
         connectUsingConfiguration(config);
         // Automatically makes the login if the user was previouslly connected successfully
         // to the server and the connection was terminated abruptly
@@ -969,7 +962,7 @@
                 }
                 else {
                     login(config.getUsername(), config.getPassword(),
+                            bind ? config.getResource() : null);
-                            config.getResource());
                 }
                 packetReader.notifyReconnection();
             }
--- org/jivesoftware/smack/PacketReader.java	2011-09-07 16:52:06.259428297 -0700
+++ org/jivesoftware/smack/PacketReader.java	2012-02-28 14:57:26.714749400 -0800
@@ -315,6 +315,14 @@
                         // to be sent by the server
                         resetParser();
                     }
+                    else {
+                    	try {
+                    		UnknownPacket packet = (UnknownPacket) PacketParserUtils.parsePacketExtension(parser.getName(), parser.getNamespace(), parser);
+                    		processPacket(packet);
+                    	} catch (ClassCastException ex) {
+                    		// ignore
+                    	}
+                    }
                 }
                 else if (eventType == XmlPullParser.END_TAG) {
                     if (parser.getName().equals("stream")) {
@@ -410,6 +401,14 @@
                 else if (parser.getName().equals("register")) {
                     connection.getAccountManager().setSupportsAccountCreation(true);
                 }
+                else {
+                	try {
+                		UnknownPacket packet = (UnknownPacket) PacketParserUtils.parsePacketExtension(parser.getName(), parser.getNamespace(), parser);
+                		processPacket(packet);
+                	} catch (ClassCastException ex) {
+                		// ignore
+                	}
+                }
             }
             else if (eventType == XmlPullParser.END_TAG) {
                 if (parser.getName().equals("starttls")) {
--- org/jivesoftware/smack/Roster.java	2012-02-28 15:23:59.774751165 -0800
+++ org/jivesoftware/smack/Roster.java	2011-09-12 20:42:45.475669892 -0700
@@ -68,6 +68,8 @@
 
     private SubscriptionMode subscriptionMode = getDefaultSubscriptionMode();
 
+	private boolean mOfflineOnError = true;
+
     /**
      * Returns the default subscription processing mode to use when a new Roster is created. The
      * subscription processing mode dictates what action Smack will take when subscription
@@ -122,7 +124,8 @@
 
             public void connectionClosedOnError(Exception e) {
                 // Changes the presence available contacts to unavailable
-                setOfflinePresences();
+            	if (mOfflineOnError)
+            		setOfflinePresences();
             }
 
         };
@@ -172,7 +175,11 @@
     public void setSubscriptionMode(SubscriptionMode subscriptionMode) {
         this.subscriptionMode = subscriptionMode;
     }
-
+    
+    public void setOfflineOnError(boolean offlineOnError) {
+		this.mOfflineOnError = offlineOnError;
+	}
+    
     /**
      * Reloads the entire roster from the server. This is an asynchronous operation,
      * which means the method will return immediately, and the roster will be
@@ -618,7 +625,7 @@
      * presence sent from the server. After a disconnection, every Presence is set
      * to offline.
      */
-    private void setOfflinePresences() {
+    public void setOfflinePresences() {
         Presence packetUnavailable;
         for (String user : presenceMap.keySet()) {
             Map<String, Presence> resources = presenceMap.get(user);
